homeassistant/components/mill/coordinator.py

```diff
@@ -40,3 +52,105 @@ def __init__(
             update_method=mill_data_connection.fetch_heater_and_sensor_data,
             update_interval=update_interval,
         )
+
+
+class MillHistoricDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching Mill historic data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        mill_data_connection: Mill,
+    ) -> None:
+        """Initialize global Mill data updater."""
+        self.mill_data_connection = mill_data_connection
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="MillHistoricDataUpdateCoordinator",
+        )
+
+    async def _async_update_data(self):
+        """Update historic data via API."""
+        now = dt_util.utcnow()
+        self.update_interval = (
+            timedelta(hours=1) + now.replace(minute=0, second=0) - now
+        )
+
+        for dev_id, heater in self.mill_data_connection.devices.items():
+            if not isinstance(heater, Heater):
+                continue
+            statistic_id = f"{DOMAIN}:energy_{slugify(dev_id)}"
+
+            last_stats = await get_instance(self.hass).async_add_executor_job(
+                get_last_statistics, self.hass, 1, statistic_id, True, set()
+            )
+
+            if not last_stats or not last_stats.get(statistic_id):
+                hourly_data = (
+                    await self.mill_data_connection.fetch_historic_energy_usage(
+                        dev_id, n_days=TWO_YEARS
+                    )
+                )
+                hourly_data = dict(sorted(hourly_data.items(), key=lambda x: x[0]))
+                _sum = 0.0
+                last_stats_time = None
+            else:
+                hourly_data = (
+                    await self.mill_data_connection.fetch_historic_energy_usage(
+                        dev_id,
+                        n_days=(
+                            now
+                            - dt_util.utc_from_timestamp(
+                                last_stats[statistic_id][0]["start"]
+                            )
+                        ).days
+                        + 2,
+                    )
+                )
+                if not hourly_data:
+                    continue
+                hourly_data = dict(sorted(hourly_data.items(), key=lambda x: x[0]))
+                start_time = next(iter(hourly_data))
+
+                stats = await get_instance(self.hass).async_add_executor_job(
```

Comment:
We called `get_last_statistics` above, how come we need to get statistics again here?

Reply:
Thanks for reviewing.

First I want to get last stored statistics, so I know from where we need to fetch data from Mill.
Then we need the last statistics from the start date of the new data from Mill, to sum the value correctly.

Sometimes it seems that data from Mill can change back in time, so I want to replace the last 2 days with fresh data from Mill.

Maybe there is a better way of doing this?

Reply:
OK, it's clear 👍 

Other integrations which import statistics follow the same pattern.
Maybe the recorder should have a function to get the first valid row at or after a start time, but that's for another PR.